include/seqan3/core/algorithm/configurator.hpp

```diff
@@ -0,0 +1,596 @@
+ * configuration type cannot occur more than once in the type definition of the configurator class.
+ *
+ * ### Usage
+ *
+ * The entire configuration system is designed to work completely in the background of any algorithm. The implementor
+ * of an algorithm only needs to provide the configurations for the algorithm.
+ * In general the type of a configuration is static (see seqan3::detail::config_base) within the context of an algorithm
+ * to select the correct code branches based on a policy-driven design, albeit the stored state might still be a runtime
+ * parameter.
+ * However, in some cases a specific configuration might be known first at runtime but needs to be converted to a static
+ * type for the algorithm in use. To enable a transparent conversion from the runtime parameters to a static type
+ * configuration one can use deferred configs (seqan3::detail::deferred_config_base), which are invocable
+ * configurations, that store the runtime parameter and on invocation translate this runtime parameter to a static
+ * configuration. The implementor of an algorithm can achieve this by using the function
+ * seqan3::detail::apply_deferred_configs, which iterates through all configurations and in case
+ * it is a deferred configuration it will invoke the translation function and continue with the modified configurator,
+ * which contains now the static type for the specific configuration.
+ *
+ * ### Pipe Notation
+ *
+ * To enable simple extension of configurations the configurator provides a generic pipe interface for the different
+ * configurations. Thus, a config class can be easily constructed by chaining together different properties.
+ * Consider the following example, where we assume that the config `bar` and the config functor `with_foo` are already
+ * given:
+ *
+ * ```cpp
+ * auto my_cfg = configurator<bar>{} | with_foo;  // my_cfg is now of type configurator<foo, bar>
+ * ```
+ * The `with_foo` functor captures the `configurator<bar>` object and returns a new configurator with the added
+ * `foo`-config.
+
+ * ### Accessor
+ *
+ * The configurator exposes a tuple interface. Thus, one can either use `seqan3::get` or `std::get` to query
+ * the state of a specific config within this configurator. Considering the example from above one can get the state of
+ * the bar-config either as:
+ *
+ * ```cpp
+ * auto bar_state = get<1>(my_cfg);
+ * ```
+ *
+ * or as:
+ *
+ * ```cpp
+ * auto bar_state = get<bar>(my_cfg);
+ * ```
+ *
+ * In addition every algorithm specific config implementation can overload the get-interface with a enum based
+ * get-interface, such that the correct config is extracted based on some type-independent enum value.
+ * This is necessary for the algorithm implementation, as the algorithm neither knows the order nor the exact type of
+ * all configs (the config type could be a template class).
+ */
```

Comment:
a) `configurator`, because it indicates some handling of stored configurations. Though, for the name of the entities it could be renamed to config_element. That is fine with me.
b) I like the design idea and will change this.
c) This is one artefact of the current implementation. At some places we want to constraint the generic type to be a configurator and therefor I first introduced it as a private concept. But giving it more thoughts, it is nothing that other libraries need to interface with really as it is only needed for our specific algorithms and thus does not need a full fledged concept definition. 
I will clean this up and use a type_trait MF instead to check if it is the correct type.